server/src/worker/import/parser.ts

```diff
@@ -18,168 +19,209 @@ const headerNames = {
-  type: "event";
-  data: {
-    id: string,
-    name: string
-  };
-}
-
-export class ImportParser extends EventEmitter {
-
-  modified = false;
-
-  result = {};
-
-  eventReader: Transform;
-  dataReader: Transform;
-
-  readable: NodeJS.ReadWriteStream;
-
-  error: Error;
-
-  constructor() {
-
-    super();
-
-    this.dataReader = this.createDataReader();
-
-    this.eventReader = this.createEventsReader();
-
-    this.readable = mergeStream(this.dataReader, this.eventReader)
-  }
-
-  async parseImport(files) {
-
-    if (files.eventsFile) {
-      var eventsFile = fs.createReadStream(files.eventsFile);
-      await this.parseEvents(eventsFile, this.eventReader);
-    }
-
-    if (files.dataFile) {
-      var dataFile = fs.createReadStream(files.dataFile);
-      await this.parseData(dataFile, this.dataReader);
+const mandatoryAccountingHeaders = [
+  "type",
+  "paragraph",
+  "item",
+  "event",
+  "amount"
+]
+
+const mandatoryPaymentsHeaders = [
+  "type",
+  "paragraph",
+  "item",
+  "unit",
+  "event",
+  "amount",
+  "date",
+  "counterpartyId",
+  "counterpartyName",
+  "description"
+]
+
+const mandatoryEventHeaders = [
+  "id",
+  "name"
+]
+
+export class ImportParser {
+
+  static createCsvParser(file: string): csvparse.Parser {
+    let headers = []
+    switch (file) {
+      case "accountingFile":
+        headers = mandatoryAccountingHeaders
+        break
+      case "paymentsFile":
+      case "dataFile":
+        headers = mandatoryPaymentsHeaders
+        break
+      case "eventsFile":
+        headers = mandatoryEventHeaders
+        break
+      default:
+        throw Error("Unexpected parser requested")
     }
 
+    return csvparse({ delimiter: ";", columns: line => this.parseHeader(line, headers), relax_column_count: true });
   }
 
-  async parseData(dataFile, reader) {
-    return new Promise((resolve, reject) => {
-
-      var error = false;
-
-      var parser = csvparse({ delimiter: ";", columns: line => this.parseHeader(line, headerNames), relax_column_count: true });
-      parser.on("error", err => (error = true, reject(err)));
-      parser.on("end", () => !error && resolve());
-
-      dataFile.pipe(parser).pipe(reader);
-    });
-  }
-
-  async parseEvents(eventsFile, reader: Readable) {
-    return new Promise((resolve, reject) => {
-
-      var error = false;
-
-      var parser = csvparse({ delimiter: ";", columns: line => this.parseHeader(line, eventHeaderNames), relax_column_count: true });
-      parser.on("error", err => (error = true, reject(err)));
-      parser.on("end", () => !error && resolve());
-
-      eventsFile.pipe(parser).pipe(reader);
-    });
+  static createCsvTransformer(file: string, options: Importer.Options): Transform {
+    switch (file) {
+      case "paymentsFile":
+      case "dataFile":
+      case "accountingFile":
+        return this.createDataTransformer(options)
+      case "eventsFile":
+        return this.createEventsReader(options)

```

Comment:
Naming unclear is it a reader or a transformer?

Reply:
Nice catch; fixed.